### PR TITLE
Add a task to monitor for users missing MFA

### PR DESCRIPTION
**Pull request by MancunianSam**
There is an existing scheduler within Keycloak which you can use. On the postInit method, we wait for the post migration event. I'm not completely sure why this event but it works. I got the idea from [this stackoverflow post](https://stackoverflow.com/questions/69656544/run-method-after-keycloak-server-startup)

We then create a new instance of the UserMonitoringTask which extends ScheduledTask and carries out the checks against the users for all realms.

If it finds users with no MFA, it sends a message to the notifications SNS topic.

### CODE REVIEW

File: event-publisher-spi/src/main/scala/uk/gov/nationalarchives/eventpublisherspi/EventPublisherProviderFactory.scala [hunk 1]
```diff
@@ -21,7 +23,16 @@ class EventPublisherProviderFactory extends EventListenerProviderFactory {
     eventPublisherConfig = Option(EventPublisherConfig(snsUrl, snsTopicArn, tdrEnvironment))
   }
 
-  override def postInit(factory: KeycloakSessionFactory): Unit = { }
+  override def postInit(factory: KeycloakSessionFactory): Unit = {
+    factory.register(event => {
+      if(event.isInstanceOf[PostMigrationEvent]) {
+         val session = factory.create()
```
Comment (TomJKing) [inline event-publisher-spi/src/main/scala/uk/gov/nationalarchives/eventpublisherspi/EventPublisherProviderFactory.scala:29]: Extra space indentation

File: event-publisher-spi/src/main/scala/uk/gov/nationalarchives/eventpublisherspi/EventPublisherProviderFactory.scala [hunk 2]
```diff
@@ -21,7 +23,16 @@ class EventPublisherProviderFactory extends EventListenerProviderFactory {
...
+        val provider: TimerProvider = session.getProvider(classOf[TimerProvider])
+        val task = UserMonitoringTask(eventPublisherConfig.get)
+        provider.scheduleTask(task, 24 * 60 * 60 * 1000, "test")
```
Comment (TomJKing) [inline event-publisher-spi/src/main/scala/uk/gov/nationalarchives/eventpublisherspi/EventPublisherProviderFactory.scala:32]: Should the `taskName` be `test`? Is it worth pulling out the timer into a val, and giving it a name to make it clearer that the unit is milliseconds?

File: event-publisher-spi/src/main/scala/uk/gov/nationalarchives/eventpublisherspi/EventPublisherProviderFactory.scala [hunk 3]
```diff
@@ -21,7 +23,16 @@ class EventPublisherProviderFactory extends EventListenerProviderFactory {
     eventPublisherConfig = Option(EventPublisherConfig(snsUrl, snsTopicArn, tdrEnvironment))
   }
 
-  override def postInit(factory: KeycloakSessionFactory): Unit = { }
+  override def postInit(factory: KeycloakSessionFactory): Unit = {
+    factory.register(event => {
+      if(event.isInstanceOf[PostMigrationEvent]) {
+         val session = factory.create()
+        val provider: TimerProvider = session.getProvider(classOf[TimerProvider])
+        val task = UserMonitoringTask(eventPublisherConfig.get)
```
Comment (TomJKing) [inline event-publisher-spi/src/main/scala/uk/gov/nationalarchives/eventpublisherspi/EventPublisherProviderFactory.scala:31]: Cold the task val be renamed: userMonitoringTask just to make it clear what task it is? Just thinking about when we come to add more tasks to run.

File: event-publisher-spi/src/main/scala/uk/gov/nationalarchives/eventpublisherspi/UserMonitoringTask.scala [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+             |The following users have missing MFA in realm $realmName
+             |${userIds.mkString("\n")}
+             |""".stripMargin)
+
+        val eventDetails = EventDetails(config.tdrEnvironment, message).asJson.toString()
+        snsUtils.publish(eventDetails, config.snsTopicArn)
+      }
+    })
+  }
+}
```
Comment (TomJKing) [inline event-publisher-spi/src/main/scala/uk/gov/nationalarchives/eventpublisherspi/UserMonitoringTask.scala:50]: Line space needed

File: event-publisher-spi/src/main/scala/uk/gov/nationalarchives/eventpublisherspi/UserMonitoringTask.scala [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+object UserMonitoringTask {
+  private val httpClient = ApacheHttpClient.builder.build
+
+  def apply(config: EventPublisherConfig): UserMonitoringTask = {
+    val snsUtils: SNSUtils = SNSUtils(SnsClient.builder()
+      .region(Region.EU_WEST_2)
+      .endpointOverride(URI.create(config.sqsUrl))
+      .httpClient(httpClient)
+      .build())
+    new UserMonitoringTask(snsUtils, config, "otp")
```
Comment (TomJKing) [inline event-publisher-spi/src/main/scala/uk/gov/nationalarchives/eventpublisherspi/UserMonitoringTask.scala:60]: Is it worth defining `otp` as a val credentialType, just to make it clear what it is?

File: event-publisher-spi/src/test/scala/uk.gov.nationalarchives.eventpublisherspi/UserMonitoringTaskSpec.scala [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+class UserMonitoringTaskSpec extends AnyFlatSpec with Matchers with MockitoSugar {
+  "The run method" should "not send a message if there are no users missing MFA" in {
+    val mockSession = mock[KeycloakSession]
+    val mockRealmProvider = mock[RealmProvider]
+    val mockRealmModel = mock[RealmModel]
+    val mockUserCredentialManager = mock[UserCredentialManager]
+    val mockUserModel = mock[UserModel]
+    val mockUserProvider = mock[UserProvider]
+    val mockSnsUtils = mock[SNSUtils]
+    val credentialType = "otp"
```
Comment (TomJKing) [inline event-publisher-spi/src/test/scala/uk.gov.nationalarchives.eventpublisherspi/UserMonitoringTaskSpec.scala:19]: Used in multiple test, maybe abstract out to a class val?

File: event-publisher-spi/src/test/scala/uk.gov.nationalarchives.eventpublisherspi/UserMonitoringTaskSpec.scala [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+    val mockSession = mock[KeycloakSession]
+    val mockRealmProvider = mock[RealmProvider]
+    val mockRealmModel = mock[RealmModel]
+    val mockUserCredentialManager = mock[UserCredentialManager]
+    val mockUserModelWithMFA = mock[UserModel]
+    val mockUserModelWithoutMFA = mock[UserModel]
+    val mockUserProvider = mock[UserProvider]
+    val mockSnsUtils = mock[SNSUtils]
+    val userId = "dfd7356c-e6e0-40dd-affd-de04e29ad359"
+    val topicArn = "snsTopicArn"
```
Comment (TomJKing) [inline event-publisher-spi/src/test/scala/uk.gov.nationalarchives.eventpublisherspi/UserMonitoringTaskSpec.scala:43]: Used in multiple tests. Is it worth abstracting out into a class val?

File: event-publisher-spi/src/test/scala/uk.gov.nationalarchives.eventpublisherspi/UserMonitoringTaskSpec.scala [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+  "The run method" should "send a message if there are two users with MFA missing" in {
+    val mockSession = mock[KeycloakSession]
+    val mockRealmProvider = mock[RealmProvider]
+    val mockRealmModel = mock[RealmModel]
+    val mockUserCredentialManager = mock[UserCredentialManager]
+    val mockUserModelWithoutMFA1 = mock[UserModel]
+    val mockUserModelWithoutMFA2 = mock[UserModel]
+    val mockUserProvider = mock[UserProvider]
+    val mockSnsUtils = mock[SNSUtils]
+    val userId1 = "dfd7356c-e6e0-40dd-affd-de04e29ad359"
```
Comment (TomJKing) [inline event-publisher-spi/src/test/scala/uk.gov.nationalarchives.eventpublisherspi/UserMonitoringTaskSpec.scala:74]: User ids used in multiple tests. Are they worth abstracting out into class vals?

**Review by TomJKing — COMMENTED** [review]
Looks good. Just some minor comments / suggestions.

If we want to add additional tasks, would that mean adding them to the `postInit` function, eg:
```
...
provider.scheduleTask(task, ... )
provider.scheduleTask(anotherTask, ...)
....
```
We would then create a new class which defines the whatever that task is? Jsut thinking ahead about other stuff we're probably going to want to monitor.

**Comment by MancunianSam** [comment]
You're right, we would create a new class extending ScheduledTask and then add in another line in the postInit function. The documentation for Keycloak is sadly lacking so I think this is the only way to do it.

File: event-publisher-spi/src/main/scala/uk/gov/nationalarchives/eventpublisherspi/EventPublisherProviderFactory.scala (hunk 2 above)
Comment (MancunianSam) [reply event-publisher-spi/src/main/scala/uk/gov/nationalarchives/eventpublisherspi/EventPublisherProviderFactory.scala:32]: I didn't mean to leave it as test :smile: I've renamed it and moved the interval out to a variable.

File: event-publisher-spi/src/main/scala/uk/gov/nationalarchives/eventpublisherspi/EventPublisherProviderFactory.scala (hunk 3 above)
Comment (MancunianSam) [reply event-publisher-spi/src/main/scala/uk/gov/nationalarchives/eventpublisherspi/EventPublisherProviderFactory.scala:31]: I've renamed it to userMonitoringTask

File: event-publisher-spi/src/main/scala/uk/gov/nationalarchives/eventpublisherspi/UserMonitoringTask.scala (hunk 5 above)
Comment (MancunianSam) [reply event-publisher-spi/src/main/scala/uk/gov/nationalarchives/eventpublisherspi/UserMonitoringTask.scala:60]: Yeah makes sense, I've moved it to a val.

File: event-publisher-spi/src/test/scala/uk.gov.nationalarchives.eventpublisherspi/UserMonitoringTaskSpec.scala (hunk 8 above)
Comment (MancunianSam) [reply event-publisher-spi/src/test/scala/uk.gov.nationalarchives.eventpublisherspi/UserMonitoringTaskSpec.scala:74]: I've moved all the user IDs as well as the topic arn and the credential type.

**Comment by TomJKing** [comment]
Have you pushed your changes? Not seeing them?

**Comment by MancunianSam** [comment]
Nope, hadn't even committed them. I've pushed them now.

**Review by TomJKing — APPROVED** [review]

**Pull request merged**